src/rgw/rgw_cache.h

```diff
@@ -216,4 +361,219 @@ class ObjectCache {
+                                 RGWObjState *astate, void *arg) {
+
+  librados::ObjectReadOperation op;
+  struct get_obj_data* d = static_cast<struct get_obj_data*>(arg);
+  string oid, key;
+  bufferlist *pbl;
+  librados::AioCompletion *c;
+
+  int r = 0;
+
+  if (is_head_obj) {
+    // only when reading from the head object do we need to do the atomic test
+    r = T::append_atomic_test(astate, op);
+    if (r < 0)
+      return r;
+
+    if (astate && obj_ofs < astate->data.length()) {
+      unsigned chunk_len = std::min((uint64_t)astate->data.length() - obj_ofs, (uint64_t)len);
+
+      r = d->client_cb->handle_data(astate->data, obj_ofs, chunk_len);
+      if (r < 0)
+        return r;
+
+      d->lock.lock();
+      d->total_read += chunk_len;
+      d->lock.unlock();
+
+      len -= chunk_len;
+      read_ofs += chunk_len;
+      obj_ofs += chunk_len;
+      if (!len)
+        return 0;
+    }
+  }
+
+  d->throttle.get(len);
+  if (d->is_cancelled()) {
+    return d->get_err_code();
+  }
+
+  // add io after we check that we're not cancelled, otherwise we're going to have trouble
+  // cleaning up
+  d->add_io(obj_ofs, len, &pbl, &c);
+
+  mydout(20) << "rados->get_obj_iterate_cb oid=" << read_obj.oid << " obj-ofs=" << obj_ofs << " read_ofs=" << read_ofs << " len=" << len << dendl;
+  op.read(read_ofs, len, pbl, NULL);
+
+  librados::IoCtx io_ctx(d->io_ctx);
+  io_ctx.locator_set_key(read_obj.loc);
+  d->add_pending_oid(read_obj.oid);
+
+  if (data_cache.get(read_obj.oid)) {
+    L1CacheRequest* cc;
+    d->add_l1_request(&cc, pbl, read_obj.oid, len, obj_ofs, read_ofs, key, c);
+    r = io_ctx.cache_aio_notifier(read_obj.oid, static_cast<CacheRequest*>(cc));
+    r = d->submit_l1_aio_read(cc);
+    if (r != 0 ){
+      mydout(0) << "Error cache_aio_read failed err=" << r << dendl;
+    }
+  } else if (d->deterministic_hash_is_local(read_obj.oid)){
+    mydout(20) << "rados->get_obj_iterate_cb oid=" << read_obj.oid << " obj-ofs=" << obj_ofs << " read_ofs=" << read_ofs << " len=" << len << dendl;
+    op.read(read_ofs, len, pbl, NULL);
+    r = io_ctx.aio_operate(read_obj.oid, c, &op, NULL);
+    mydout(20) << "rados->aio_operate r=" << r << " bl.length=" << pbl->length() << dendl;
+    if (r < 0) {
+      mydout(0) << "rados->aio_operate r=" << r << dendl;
+      goto done_err;
+    }
+  }  else {
+    L2CacheRequest* cc;
+    d->add_l2_request(&cc, pbl, read_obj.oid, obj_ofs, read_ofs, len, key, c);
+    r = io_ctx.cache_aio_notifier(read_obj.oid, static_cast<CacheRequest*>(cc));
+    data_cache.push_l2_request(cc);
+  }
+
+  // Flush data to client if there is any
+  r = flush_read_list(d);
+  if (r < 0)
+    return r;
+
+  return 0;
+
+done_err:
+  mydout(20) << "cancelling io r=" << r << " obj_ofs=" << obj_ofs << dendl;
+  d->set_cancelled(r);
+  d->cancel_io(obj_ofs);
+
+  return r;
+}
+
+class L2CacheThreadPool {
+public:
+  L2CacheThreadPool(int n) {
+    curl_global_init(CURL_GLOBAL_ALL);
```

Comment:
rgw_http_client.cc is already calling this

Reply:
checked the curl_global_init call order, this one is redundant, will remove.